File: Design-Patterns-Using-Ruby/Momento/MomentoPattern.rb
# Memento
class Memento
    attr_reader :state

    def initialize(state)
        @state = state
    end
end

# CareTaker
class CareTaker
    attr_reader :history, :top, :max

    def initialize
        @history = []
        @top = -1
        @max = -1
    end

    def add_memento(memento)
        @top += 1
        @max = @top

        if @top <= @history.size - 1
            @history[@top] = memento
        else
            @history << memento
        end
    end

    def get_memento(index)
        @history[index]
    end

    def undo
        puts 'Undoing state.'
        if @top <= 0
            @top = 0
            return get_memento(0)
        end

        @top -= 1
        get_memento(@top)
    end

    def redo
        puts 'Redoing state.'
        if @top >= (@history.size - 1) || @top >= @max
            return get_memento(@top)
        end

        @top += 1
        get_memento(@top)
    end

    def states_count
        @history.size
    end
end

# Originator
class Originator
    attr_accessor :state

    def set_state(state)
        @state = state
    end
    
    def get_state
        @state
    end

    def create_memento
        Memento.new(state)
    end

    def set_memento(memento)
        self.state = memento.state
    end
end

# Client code
originator = Originator.new
care_taker = CareTaker.new

originator.set_state('State 1')
care_taker.add_memento(originator.create_memento)
puts originator.state

originator.set_state('State 2')
care_taker.add_memento(originator.create_memento)
puts originator.state

originator.set_state('State 3')
care_taker.add_memento(originator.create_memento)
puts originator.state

originator.set_memento(care_taker.undo)
puts originator.state

originator.set_memento(care_taker.undo)
puts originator.state

originator.set_memento(care_taker.redo)
puts originator.state

originator.set_memento(care_taker.redo)
puts originator.state

=begin 
State 1
State 2
State 3
Undoing state.
State 2
Undoing state.
State 1
Redoing state.
State 2
Redoing state.
State 3
=end

File: Design-Patterns-Using-Kotlin/Momento/MomentoPattern.kt
// Memento
class Memento(val state: String?)

// CareTaker
class CareTaker {
    private val history: MutableList<Memento> = ArrayList()
    private var top: Int = -1
    private var max: Int= -1

    fun addMemento(m: Memento) {
        top += 1
        max = top
        if (top <= history.size - 1) {
            history[top] = m
        } else {
            history.add(m)
        }
    }

    fun getMemento(index: Int): Memento {
        return history[index]
    }

    fun undo(): Memento {
        println("Undoing state.")
        if (top <= 0) {
            top = 0
            return getMemento(0)
        }

        top -= 1
        return getMemento(top)
    }

    fun redo(): Memento {
        println("Redoing state.")
        if (top >= (history.size - 1) || top >= max) {
            return getMemento(top)
        }

        top += 1
        return getMemento(top)
    }

    val statesCount: Int
        get() = history.size
}

// Originator
class Originator {
    var state: String? = null

    fun createMemento(): Memento {
        return Memento(state)
    }

    fun setMemento(m: Memento) {
        state = m.state
    }
}

// Client code
fun main() {
    val originator = Originator()
    val careTaker = CareTaker()

    originator.state = "State 1"
    careTaker.addMemento(originator.createMemento())
    println(originator.state)

    originator.state = "State 2"
    careTaker.addMemento(originator.createMemento())
    println(originator.state)

    originator.state = "State 3"
    careTaker.addMemento(originator.createMemento())
    println(originator.state)

    originator.setMemento(careTaker.undo())
    println(originator.state)

    originator.setMemento(careTaker.undo())
    println(originator.state)

    originator.setMemento(careTaker.redo())
    println(originator.state)

    originator.setMemento(careTaker.redo())
    println(originator.state)
}

/*
State 1
State 2
State 3
Undoing state.
State 2
Undoing state.
State 1
Redoing state.
State 2
Redoing state.
State 3
*/